### PR TITLE
kubeadm: Add configurable control plane up timeout

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Until now the control plane timeout was fixed to 4 minutes and users did not have the ability to change it. This commit allows that timeout to be configured via the new `timeoutForControlPlane` option in the API server config (itself a member of the ClusterConfiguration).
    
The default timeout is still 4 minutes.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Refs kubernetes/kubeadm#911

**Special notes for your reviewer**:

This PR depends on:
- [X] #70371 
- [X] #70495 

Please, review only the last commit here.

This PR does not modify the self hosting timeout code, which is going to be deleted in another PR:
- [ ] #69878 

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/assign @fabriziopandini 
/assign @timothysc

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: timeoutForControlPlane is introduced as part of the API Server config, that controls the timeout for the wait for control plane to be up. Default value is 4 minutes.
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go]: should we use the default here 4m?

**Review by neolit123 — APPROVED** [review]
thanks @rosti

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go]: I don't think it's necessary.

Comment (timothysc) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/doc.go]: I agree with @neolit123, please change to 4 b/c that's the default used all over.

**Review by timothysc — APPROVED** [review]
/approve 

Generally looks like what we talked about, thx for the patch.

**Review by fabriziopandini — APPROVED** [review]
Thanks @rosti!
/approve
Waiting for dependency to merge before final lgtm

**Comment by rosti** [comment]
Hold because of unmerged dependency PR.

/hold

**Comment by neolit123** [comment]
@rosti @fabriziopandini 

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/apiclient/wait.go#L159
^ it runs a couple of go routines concurrently an whichever fails first returns an error. so if f() there is for instance WaitForAPI() (for init / wait-control-plane phase) the control plane timeout would be respected but the other target is always WaitForHealthyKubelet() which has a timeout of total 195 seconds. so even if the user sets DefaultControlPlaneTimeout to 10 minutes it will always fail in no more than 195 seconds.

**Comment by fabriziopandini** [comment]
@neolit123 @rosti 
In my opionin it makes sense
- if kubelet doesn't start, fail fast after 195 sec (does not make sense wait for the control plane)
– if kubelet starts, wait for control plane max 4 minutes

**Comment by neolit123** [comment]
i think what has to be done here is not use exponential backoff in WaitForHealthyKubelet, but rather allow us to make the function match a timeout the user has specified - e..g. DefaultKubeletTimeout

File: cmd/kubeadm/app/cmd/init.go
Comment (dixudx) [inline cmd/kubeadm/app/cmd/init.go]: Why declaring this var, which is used only once?

**Review by dixudx — APPROVED** [review]

File: cmd/kubeadm/app/cmd/init.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/init.go]: That's because the expression to take the timeout value is huge and best viewed on a single line. Also the call to `NewControlPlaneWaiter` on the following line continues to fit on screen.

**Comment by rosti** [comment]
@neolit123 I agree with you, that the waiter code can be simplified a bit (to say the least), but perhaps this is a job for another PR?

**Review by fabriziopandini — APPROVED** [review]
@rosti Thanks for this PR!
/approve
/lgtm

@neolit123 @dixudx we can eventually iterate on details, but I'm moving on with this PR to get all the API changes in before code slush

**Pull request merged**
